dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/EndpointErrorDecoder.java

```diff
@@ -0,0 +1,268 @@
+                }
+                break;
+            case 429:
+                return Optional.of(response.getFirstHeader(HttpHeaders.RETRY_AFTER)
+                        .map(Longs::tryParse)
+                        .map(Duration::ofSeconds)
+                        .map(duration -> QosException.throttle(qosReason(response), duration))
+                        .orElseGet(() -> QosException.throttle(qosReason(response))));
+            case 503:
+                return Optional.of(QosException.unavailable(qosReason(response)));
+        }
+        return Optional.empty();
+    }
+
+    private @Nullable String extractErrorName(byte[] body) {
+        try {
+            NamedError namedError = NAMED_ERROR_DESERIALIZER.deserialize(new ByteArrayInputStream(body));
+            if (namedError == null) {
+                return null;
+            }
+            return namedError.errorName();
+        } catch (IOException | RuntimeException e) {
+            return null;
+        }
+    }
+
+    private T decodeInternal(Response response) {
+        Optional<RuntimeException> maybeQosException = checkCode(response);
+        if (maybeQosException.isPresent()) {
+            throw maybeQosException.get();
+        }
+        int code = response.code();
+
+        byte[] body;
+        try {
+            body = toByteArray(response.body());
+        } catch (NullPointerException | IOException e) {
+            UnknownRemoteException exception = new UnknownRemoteException(code, "<unparseable>");
+            exception.initCause(e);
+            throw exception;
+        }
+
+        Optional<String> contentType = response.getFirstHeader(HttpHeaders.CONTENT_TYPE);
+        if (contentType.isPresent()
+                && Encodings.matchesContentType(JSON_ENCODING.getContentType(), contentType.get())) {
+            try {
+                String errorName = extractErrorName(body);
+                if (errorName == null) {
+                    throw createRemoteException(body, code);
+                }
+                Deserializer<? extends T> deserializer = errorNameToJsonDeserializerMap.get(errorName);
+                if (deserializer == null) {
+                    throw createRemoteException(body, code);
+                }
+                return deserializer.deserialize(new ByteArrayInputStream(body));
+            } catch (RemoteException remoteException) {
+                // rethrow the created remote exception
+                throw remoteException;
+            } catch (Exception e) {
+                throw new UnknownRemoteException(code, new String(body, StandardCharsets.UTF_8));
+            }
+        }
+
+        throw new UnknownRemoteException(code, new String(body, StandardCharsets.UTF_8));
+    }
+
+    private static RemoteException createRemoteException(byte[] body, int code) throws IOException {
+        SerializableError serializableError = JSON_ENCODING
+                .deserializer(new TypeMarker<SerializableError>() {})
```

Comment:
Can we initialize this `Deserializer<SerializableError>` once above, rather than creating it anew each time it's needed?